doc/design/GeneratedClientDesign.md

```diff
@@ -0,0 +1,133 @@
+bytes | bytes[] | type: string, format: byte | [autorest.csharp #2337](https://github.com/Azure/autorest.csharp/issues/2337) |
+int32  | int | type: integer, format: int32 | [autorest.csharp #2337](https://github.com/Azure/autorest.csharp/issues/2337) |
+int64  | long | type: integer, format: int64 | [autorest.csharp #2337](https://github.com/Azure/autorest.csharp/issues/2337) |
+safeint  | long | n/a | [autorest.csharp #2337](https://github.com/Azure/autorest.csharp/issues/2337) |
+float32  | float | type: number, format: float | [autorest.csharp #2337](https://github.com/Azure/autorest.csharp/issues/2337) |
+float64  | double | type: number, format: double | [autorest.csharp #2337](https://github.com/Azure/autorest.csharp/issues/2337) |
+zonedDateTime  | DateTimeOffset | type: string, format: date-time | [autorest.csharp #2337](https://github.com/Azure/autorest.csharp/issues/2337) | Serialized differently based on body/header
+duration  | TimeSpan | type: string, format: duration | [autorest.csharp #2337](https://github.com/Azure/autorest.csharp/issues/2337) |
+boolean  | boolean | boolean | [autorest.csharp #2337](https://github.com/Azure/autorest.csharp/issues/2337) |
+<!-- TBD | TBD | type: string, format: binary | [autorest.csharp #2500](https://github.com/Azure/autorest.csharp/issues/2500)| -->
+
+### Model Shape
+
+Model APIs will differ depending on whether the model is an input model, an output model, or a round-trip model (both input and output).
+
+The following describes the model shapes for different cases of properties.
+
+ Item |  Input | Output | Round-Trip | GitHub Issue
+-- | -------| ------------ | -------- | -- |
+Constructor Accessibility | public | internal | public |
+Required Property | get-only | get-only | get/set | [autorest.csharp #2463](https://github.com/Azure/autorest.csharp/issues/2463)
+Optional Property | get/set | get-only | get/set | [autorest.csharp #2339](https://github.com/Azure/autorest.csharp/issues/2339)
+Collection Property | `IList<T>` get-only | `IReadOnlyList<T>` get-only | `IList<T>` get-only | [autorest.csharp #2471](https://github.com/Azure/autorest.csharp/issues/2471)
+
+### Model Constructors
+
+Models have one or more constructors as follows.
+
+#### Main Constructor
+
+- Accessibility is specified in model shape table above
+- Takes required parameters and does not take optional parameters
+- Validates required reference type parameters for null using `Argument.AssertNotNull`
+- Takes list properties as `IEnumerable<T>` parameters
+- Initializes lists from `IEnumerable` parameter using System.Linq `.ToList()`
+
+#### Serialization Constructor
+
+- Internal accessibility
+- Generated for Output and Round-trip types
+- Takes required and optional parameters
+- Takes list properties as `IReadOnlyList<T>` parameters for Output models and `IList<T>` parameters for Round-trip models
```

Comment:
Semantics of first and second constructor are very different.  One makes a copy, the other is a reference so see changes.

This ok because it's an internal constructor used by serializer.